libnmstate/nm/device.py

```diff
@@ -60,14 +150,47 @@ def _active_connection_callback(src_object, result, user_data):
                     nmdev.get_iface(), e))
         return
 
-    if act_con is None:
+    if nm_act_con is None:
         mainloop.quit('Connection activation failed on %s: error=unknown' %
                       nmdev.get_iface())
     else:
-        devname = act_con.props.connection.get_interface_name()
+        devname = nm_act_con.props.connection.get_interface_name()
+        logging.debug('Connection activation initiated: dev=%s, con-state=%s',
+                      devname, nm_act_con.props.state)
+
+        ac = ActiveConnection(nm_act_con)
+        ac.refresh_state()
```

Comment:
IMHO it makes sense to refresh the state in the constructor or initialize with the correct state from the beginning.

Reply:
Yes it can be refresh inside the constructor.
But as the object represents a state of several underlying objects (of NM), and those states are only true at the time of "refresh", the object state is never "exact". I think it is clearer for such objects to get constructed with an init state and only at later stages (before use) to be refreshed explicitly.

If you have strong feeling on this, I can change it. Please let me know.

Reply:
Why does it need to be manually refreshed at all? How about refreshing the state whenever one of the properties is accessed?

Reply:
I just tried to change it. Looks like I need to call refresh from 4 places now: `is_active`, `is_activating`, `reason` and `state`. (for the last I could use a direct call to self._act_con.get_state(), but not for the other 3)